cardano-crypto-class/src/Cardano/Crypto/KES/CompactSum.hs

```diff
@@ -0,0 +1,400 @@
+  , CompactSum7KES
+  ) where
+
+import           Data.Proxy (Proxy(..))
+import           Data.Typeable (Typeable)
+import           GHC.Generics (Generic)
+import qualified Data.ByteString as BS
+import           Control.Monad (guard)
+import           NoThunks.Class (NoThunks)
+
+import           Cardano.Binary (FromCBOR (..), ToCBOR (..))
+
+import           Cardano.Crypto.Seed
+import           Cardano.Crypto.Hash.Class
+import           Cardano.Crypto.KES.Class
+import           Cardano.Crypto.KES.CompactSingle (CompactSingleKES)
+import Data.Word (Word8)
+import Control.DeepSeq (NFData)
+
+
+-- | A 2^0 period KES
+type CompactSum0KES d   = CompactSingleKES d
+
+-- | A 2^1 period KES
+type CompactSum1KES d h = CompactSumKES h (CompactSum0KES d)
+
+-- | A 2^2 period KES
+type CompactSum2KES d h = CompactSumKES h (CompactSum1KES d h)
+
+-- | A 2^3 period KES
+type CompactSum3KES d h = CompactSumKES h (CompactSum2KES d h)
+
+-- | A 2^4 period KES
+type CompactSum4KES d h = CompactSumKES h (CompactSum3KES d h)
+
+-- | A 2^5 period KES
+type CompactSum5KES d h = CompactSumKES h (CompactSum4KES d h)
+
+-- | A 2^6 period KES
+type CompactSum6KES d h = CompactSumKES h (CompactSum5KES d h)
+
+-- | A 2^7 period KES
+type CompactSum7KES d h = CompactSumKES h (CompactSum6KES d h)
+
+
+-- | A composition of two KES schemes to give a KES scheme with the sum of
+-- the time periods.
+--
+-- While we could do this with two independent KES schemes (i.e. two types)
+-- we only need it for two instances of the same scheme, and we save
+-- substantially on the size of the type and runtime dictionaries if we do it
+-- this way, especially when we start applying it recursively.
+--
+data CompactSumKES h d
+
+instance (NFData (SigKES d), NFData (VerKeyKES d)) =>
+  NFData (SigKES (CompactSumKES h d)) where
+
+instance (NFData (SignKeyKES d), NFData (VerKeyKES d)) =>
+  NFData (SignKeyKES (CompactSumKES h d)) where
+
+instance (OptimizedKESAlgorithm d, HashAlgorithm h, Typeable d)
+      => KESAlgorithm (CompactSumKES h d) where
+
+    type SeedSizeKES (CompactSumKES h d) = SeedSizeKES d
+
+    --
+    -- Key and signature types
+    --
+
+    -- | From Section 3,1:
+    --
+    -- The verification key @vk@ for the sum scheme is the hash of the
+    -- verification keys @vk_0, vk_1@ of the two constituent schemes.
+    --
+    newtype VerKeyKES (CompactSumKES h d) =
+              VerKeyCompactSumKES (Hash h (VerKeyKES d, VerKeyKES d))
+        deriving Generic
+        deriving newtype NFData
+
+    -- | From Figure 3: @(sk_0, r_1, vk_0, vk_1)@
+    --
+    data SignKeyKES (CompactSumKES h d) =
+           SignKeyCompactSumKES !(SignKeyKES d)
+                         !Seed
+                         !(VerKeyKES d)
+                         !(VerKeyKES d)
+        deriving Generic
+
+    -- | From Figure 3: @(sigma, vk_0, vk_1)@
+    --
+    data SigKES (CompactSumKES h d) =
+           SigCompactSumKES !(SigKES d) -- contains VerKey for this branch!
```

Comment:
Does it? The `SigKES` does not contain the `VerKey` of the whole tree. It simply contains the DSIGN verkey, right? It does contain the data to be able to compute it though.

Reply:
It is only the `CompactSingle` that contains the `VerKey` of its own branch/leaf.

Reply:
Correct. Fixing.